proto/src/rr/rdata/opt.rs

```diff
@@ -196,24 +196,31 @@ impl OPT {
 }
 
 /// Read the RData from the given Decoder
-pub fn read(decoder: &mut BinDecoder, rdata_length: u16) -> ProtoResult<OPT> {
+pub fn read(decoder: &mut BinDecoder, rdata_length: Restrict<u16>) -> ProtoResult<OPT> {
     let mut state: OptReadState = OptReadState::ReadCode;
     let mut options: HashMap<EdnsCode, EdnsOption> = HashMap::new();
     let start_idx = decoder.index();
 
-    while rdata_length as usize > decoder.index() - start_idx {
+    // There is no unsafe direct use of the rdata length after this point
+    let rdata_length = rdata_length.map(|u| u as usize).unverified();
+    while rdata_length > decoder.index() - start_idx {
         match state {
-            // TODO: can condense Code1/2 and Length1/2 into read_u16() calls.
             OptReadState::ReadCode => {
                 state = OptReadState::Code {
-                    code: (decoder.read_u16()?).into(),
+                    code: EdnsCode::from(decoder.read_u16()?.unverified()),
                 };
             }
             OptReadState::Code { code } => {
-                let length: usize = decoder.read_u16()? as usize;
+                let length = decoder
+                    .read_u16()?
+                    .map(|u| u as usize)
+                    .verify_unwrap(|u| *u <= rdata_length)
```

Comment:
double checking that the length here is in bounds, could have been an excessive read